File: content/snippets/extract-people-and-companies-from-an-article/n2.test.py
import time

import pytest

from _harness.fake_model import FakeClassifier
from n0 import extract_names as extract_names_n0
from n2 import LABELS, MAX_CHARACTERS, RecognitionUnavailable, extract_names

ARTICLE = ("Le contrat lie la société Lumière SARL à Jean de La Fontaine et à "
           "Mme Marie Martin, de Lyon. M. Boulanger a livré le colis mercredi.")


def entites(*trouvees):
    return {"entities": [{"text": t, "label": l, "start": ARTICLE.index(t),
                          "end": ARTICLE.index(t) + len(t)} for t, l in trouvees]}


def double(reponse=None, **extra):
    return FakeClassifier({ARTICLE: reponse if reponse is not None else entites(
        ("Lumière SARL", "ORG"), ("Jean de La Fontaine", "PER"),
        ("Marie Martin", "PER"), ("Lyon", "LOC"), ("Boulanger", "PER"))},
        default={"entities": []}, **extra)


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_le_modele_type_tout_et_ne_rend_aucun_doute():
    """
    « Le pipeline rend une étiquette pour chaque nom qu'il trouve, et rien à
    côté : il n'y a pas de seuil à monter ni d'« unknown » sur quoi retomber. »
    """
    lu = extract_names([ARTICLE], double(entites(("Lyon", "ORG"))))[0]
    assert lu["names"] == [{"text": "Lyon", "type": "company", "evidence": "ORG",
                            "start": ARTICLE.index("Lyon"),
                            "end": ARTICLE.index("Lyon") + 4}]
    # Rien dans le rapport ne permet d'en douter : pas de score, pas de doute.
    assert "score" not in lu["names"][0]


def test_point_de_rupture_temoin_le_niveau_n0_repond_unknown_sur_le_meme_nom():
    """
    « Le témoin est dans le même test : sur ce même nom, le niveau N0 répond
    `unknown`, parce que la phrase ne prouve rien. »
    """
    lu = extract_names_n0(ARTICLE)["names"]
    lyon = [n for n in lu if n["text"] == "Lyon"][0]
    assert (lyon["type"], lyon["evidence"]) == ("unknown", None)


# ---------------------------------------------------------------------------
# Les autres affirmations du niveau
# ---------------------------------------------------------------------------


def test_une_etiquette_non_cartographiee_est_ecartee_et_comptee():
    lu = extract_names([ARTICLE], double(entites(("Lyon", "LOC"), ("colis", "MISC"))))[0]
    assert [n["text"] for n in lu["names"]] == ["Lyon"]
    assert lu["unmapped"] == ["MISC"]


def test_les_etiquettes_des_deux_jeux_courants_sont_cartographiees():
    assert LABELS["PER"] == "person" and LABELS["PERSON"] == "person"
    assert LABELS["ORG"] == "company" and LABELS["LOC"] == LABELS["GPE"] == "place"


def test_un_nom_absent_du_texte_envoye_est_ecarte():
    """Le modèle rend une portion du texte ; ce qui n'y est pas n'en vient pas."""
    lu = extract_names([ARTICLE], FakeClassifier(
        {ARTICLE: {"entities": [{"text": "Marseille", "label": "LOC", "start": 0, "end": 9}]}},
        default={"entities": []}))[0]
    assert lu["names"] == []


def test_le_texte_est_coupe_au_budget_pas_refuse():
    long_texte = ARTICLE * 1000
    lu = extract_names([long_texte], FakeClassifier({}, default={"entities": []}))[0]
    assert lu["characters_read"] == MAX_CHARACTERS


def test_une_panne_du_modele_est_nommee():
    class Casse:
        def predict(self, texts):
            raise RuntimeError("modèle introuvable")

    with pytest.raises(RecognitionUnavailable):
        extract_names([ARTICLE], Casse())


def test_une_reponse_qui_na_pas_la_forme_attendue_est_refusee():
    class Bavard:
        def predict(self, texts):
            return [{"entities": []}, {"entities": []}]

    with pytest.raises(RecognitionUnavailable):
        extract_names([ARTICLE], Bavard())


def test_le_lot_est_envoye_en_une_fois():
    client = double()
    extract_names([ARTICLE, "Autre texte."], client)
    assert client.calls == [[ARTICLE, "Autre texte."]]


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_entree_banale_un_paragraphe_de_presse():
    """T5 : l'entrée ordinaire du public visé."""
    lu = extract_names([ARTICLE], double())[0]
    assert [(n["text"], n["type"]) for n in lu["names"]] == [
        ("Lumière SARL", "company"), ("Jean de La Fontaine", "person"),
        ("Marie Martin", "person"), ("Lyon", "place"), ("Boulanger", "person")]


def test_production_entree_vide():
    lu = extract_names([""], FakeClassifier({}, default={"entities": []}))[0]
    assert lu == {"names": [], "unmapped": [], "characters_read": 0}


def test_production_valeurs_aux_limites():
    vide = FakeClassifier({}, default={"entities": []})
    assert extract_names([], vide) == []
    # Une entrée qui n'est pas du texte devient un texte vide, pas une exception.
    assert extract_names([None, 42], vide)[0]["characters_read"] == 0
    # Une réponse sans clé « entities ».
    assert extract_names([ARTICLE], FakeClassifier({}, default={}))[0]["names"] == []


def test_production_un_texte_illisible_nempeche_pas_de_lire_les_autres():
    """T8 : un texte muet dans le lot ne fait pas tomber les autres."""
    rapports = extract_names([ARTICLE, None], double())
    assert len(rapports[0]["names"]) == 5
    assert rapports[1]["names"] == []


def test_production_la_lecture_tient_la_classe_de_latence_annoncee():
    """latency « ~100 ms » : le travail local, hors modèle, reste négligeable."""
    client = double()
    debut = time.perf_counter()
    for _ in range(500):
        extract_names([ARTICLE], client)
    assert time.perf_counter() - debut < 10.0

File: content/snippets/extract-people-and-companies-from-an-article/n2.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import { FakeClassifier } from '../_harness/fake-model.mjs';
import { extractNames as extractNamesN0 } from './n0.js';
import { LABELS, MAX_CHARACTERS, RecognitionUnavailable, extractNames } from './n2.js';

const ARTICLE = 'Le contrat lie la société Lumière SARL à Jean de La Fontaine et à '
  + 'Mme Marie Martin, de Lyon. M. Boulanger a livré le colis mercredi.';

const entites = (...trouvees) => ({
  entities: trouvees.map(([texte, label]) => ({
    text: texte,
    label,
    start: ARTICLE.indexOf(texte),
    end: ARTICLE.indexOf(texte) + texte.length,
  })),
});

const TROUVES = entites(['Lumière SARL', 'ORG'], ['Jean de La Fontaine', 'PER'],
  ['Marie Martin', 'PER'], ['Lyon', 'LOC'], ['Boulanger', 'PER']);

const double = (reponse = TROUVES) => new FakeClassifier({ [ARTICLE]: reponse }, { entities: [] });

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test("point de rupture : le modèle type tout et ne rend aucun doute", async () => {
  const lu = (await extractNames([ARTICLE], double(entites(['Lyon', 'ORG']))))[0];
  assert.deepEqual(lu.names, [{
    text: 'Lyon',
    type: 'company',
    evidence: 'ORG',
    start: ARTICLE.indexOf('Lyon'),
    end: ARTICLE.indexOf('Lyon') + 4,
  }]);
  assert.ok(!('score' in lu.names[0]));
});

test('point de rupture : témoin, le niveau N0 répond « unknown » sur le même nom', () => {
  const lyon = extractNamesN0(ARTICLE).names.find((n) => n.text === 'Lyon');
  assert.deepEqual([lyon.type, lyon.evidence], ['unknown', null]);
});

// ---------------------------------------------------------------------------
// Les autres affirmations du niveau
// ---------------------------------------------------------------------------

test('une étiquette non cartographiée est écartée et comptée', async () => {
  const lu = (await extractNames([ARTICLE],
    double(entites(['Lyon', 'LOC'], ['colis', 'MISC']))))[0];
  assert.deepEqual(lu.names.map((n) => n.text), ['Lyon']);
  assert.deepEqual(lu.unmapped, ['MISC']);
});

test('les étiquettes des deux jeux courants sont cartographiées', () => {
  assert.equal(LABELS.PER, 'person');
  assert.equal(LABELS.PERSON, 'person');
  assert.equal(LABELS.ORG, 'company');
  assert.equal(LABELS.LOC, 'place');
  assert.equal(LABELS.GPE, 'place');
});

test("un nom absent du texte envoyé est écarté", async () => {
  const menteur = new FakeClassifier(
    { [ARTICLE]: { entities: [{ text: 'Marseille', label: 'LOC', start: 0, end: 9 }] } },
    { entities: [] },
  );
  const lu = (await extractNames([ARTICLE], menteur))[0];
  assert.deepEqual(lu.names, []);
});

test('le texte est coupé au budget, pas refusé', async () => {
  const longTexte = ARTICLE.repeat(1000);
  const lu = (await extractNames([longTexte], new FakeClassifier({}, { entities: [] })))[0];
  assert.equal(lu.characters_read, MAX_CHARACTERS);
});

test('une panne du modèle est nommée', async () => {
  const casse = { predict: async () => { throw new Error('modèle introuvable'); } };
  await assert.rejects(() => extractNames([ARTICLE], casse), RecognitionUnavailable);
});

test("une réponse qui n'a pas la forme attendue est refusée", async () => {
  const bavard = { predict: async () => [{ entities: [] }, { entities: [] }] };
  await assert.rejects(() => extractNames([ARTICLE], bavard), RecognitionUnavailable);
});

test('le lot est envoyé en une fois', async () => {
  const client = double();
  await extractNames([ARTICLE, 'Autre texte.'], client);
  assert.deepEqual(client.calls, [[ARTICLE, 'Autre texte.']]);
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : entrée banale, un paragraphe de presse', async () => {
  const lu = (await extractNames([ARTICLE], double()))[0];
  assert.deepEqual(lu.names.map((n) => [n.text, n.type]), [
    ['Lumière SARL', 'company'], ['Jean de La Fontaine', 'person'],
    ['Marie Martin', 'person'], ['Lyon', 'place'], ['Boulanger', 'person'],
  ]);
});

test('production : entrée vide', async () => {
  const lu = (await extractNames([''], new FakeClassifier({}, { entities: [] })))[0];
  assert.deepEqual(lu, { names: [], unmapped: [], characters_read: 0 });
});

test('production : valeurs aux limites', async () => {
  const vide = new FakeClassifier({}, { entities: [] });
  assert.deepEqual(await extractNames([], vide), []);
  assert.equal((await extractNames([null, 42], vide))[0].characters_read, 0);
  const muet = new FakeClassifier({}, {});
  assert.deepEqual((await extractNames([ARTICLE], muet))[0].names, []);
});

test("production : un texte illisible n'empêche pas de lire les autres", async () => {
  const rapports = await extractNames([ARTICLE, null], double());
  assert.equal(rapports[0].names.length, 5);
  assert.deepEqual(rapports[1].names, []);
});

test('production : la lecture tient la classe de latence annoncée', async () => {
  const client = double();
  const debut = performance.now();
  for (let i = 0; i < 500; i += 1) await extractNames([ARTICLE], client);
  assert.ok(performance.now() - debut < 10_000);
});
